### PR TITLE
libexpr: add more precise position information to evaluation error messages

### PR DESCRIPTION
**Pull request by polykernel**
# Motivation

The current evaluation error trace messages are somewhat generic. When evaluating elements of a list or attribute set, we pass a static error context indicating one of the elements is being evaluated. For example, in the `prim_catAttrs` https://github.com/NixOS/nix/blob/27f82ef4a8fc589738ef06040dad5c0e214f97aa/src/libexpr/primops.cc#L2619-L2620 and in `EvalState::concatLists` https://github.com/NixOS/nix/blob/27f82ef4a8fc589738ef06040dad5c0e214f97aa/src/libexpr/eval.cc#L1894-L1910

Since we know the position of the value we are trying to evaluate from the caller side, we can add this information to the error context to give more helpful traces. A naive approach to add this information would be to construct a formatted hint from the caller side and passing it to the `foceValue` functions. However, this means the formatted hint is built on each call even if no error occurs which has a negative performance impact. This PR attempts to workaround this problem by altering the signature of the `forceValue` functions and forwarding the format arguments to their respective callee so that the formatted hint is only constructed when an evaluation error occurs.

# Context

Initiated from the discussion at https://github.com/NixOS/nix/pull/8398#discussion_r1206513778

Benchmarks:
I performed the benchmarks using hyperfine with the current master branch being compared to this PR.

before
```
Benchmark 1: nix eval --extra-experimental-features nix-command -f ../nixpkgs/pkgs/development/haskell-modules/hackage-packages.nix
  Time (mean ± σ):     402.8 ms ±   2.5 ms    [User: 366.8 ms, System: 36.1 ms]
  Range (min … max):   399.6 ms … 407.6 ms    20 runs

Benchmark 2: nix search --extra-experimental-features "nix-command flakes" --no-eval-cache --offline ../nixpkgs hello
  Time (mean ± σ):     17.247 s ±  0.183 s    [User: 15.192 s, System: 2.128 s]
  Range (min … max):   16.958 s … 17.538 s    20 runs
```

after
```
Benchmark 1: nix eval --extra-experimental-features nix-command -f ../nixpkgs/pkgs/development/haskell-modules/hackage-packages.nix
  Time (mean ± σ):     401.2 ms ±   3.2 ms    [User: 363.0 ms, System: 38.2 ms]
  Range (min … max):   397.2 ms … 409.5 ms    20 runs

Benchmark 2: nix search --extra-experimental-features "nix-command flakes" --no-eval-cache --offline ../nixpkgs hello
  Time (mean ± σ):     17.322 s ±  0.374 s    [User: 15.235 s, System: 2.164 s]
  Range (min … max):   16.728 s … 18.134 s    20 runs
```

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [x] agreed on implementation strategy
 - [x] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [x] code and comments are self-explanatory
 - [x] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by polykernel** [comment]
@roberth I am still contemplating whether the approach will work universally but I would like to hear your thoughts. I also wanted to ask whether API breakage for libexpr is acceptable as I had some things I wanted to implement that would change how some functions are called.

**Comment by Ericson2314** [comment]
CC @layus

**Comment by layus** [comment]
It makes sense to me, and the requirement to not slow execution by assembling error contexts on the fly has been addressed. So as long as everyone is okay with the code changes, I see no problem merging this.

Now, contexts were introduced with two main goals / advantages:
1. Forcing users to provide some context when they call a function likely to fail. (casting a value to a given type, mostly)
2. Avoid the bloat of having to wrap every call in a try-catch block.
3. Avoid allocating a fmt() object on the stack, at least in "hot" functions.

So, instead of making context _that_ more complex, with varargs and templates, would you consider wrapping the call site with a try-catch block that adds the error context "manually" ?

**Comment by Ericson2314** [comment]
@layus Are you proposing sort of the opposite approach, where we do more `try..catch` in the caller instead of less? Sounds good to me.

**Comment by layus** [comment]
@Ericson2314 Yes, that is my proposition. It depends a lot on the number of future occurrences of that pattern. I have counted 8 in this PR.

Now that I have had a chance to look at it properly, I find it quite elegant, given the context. It's a proper generalization of my static contexts. It gets a chance to inline, and then benefits from the zero-overhead of throw/catch.
If properly inlined, you can even pass computed values to it, like the type of a value.

Not sure if we can bring this pattern much further, but it seems to do the job as-is, well done !

I would love to get a perf review from someone with a deeper understanding of inlining, memory usage and code-size impact.

**Comment by polykernel** [comment]
> So, instead of making context that more complex, with varargs and templates, would you consider wrapping the call site with a try-catch block that adds the error context "manually" ?

I did initially considered following this approach but I thought adding try-catch blocks around each call site would introduce too much clutter. I think there is also the lifetime issue since objects inside the try block get destroyed at the end of the block so one needs to nest try-catch to evaluate multiple values with different context or resort to using pointers with sequential try-catch blocks.

**Comment by layus** [comment]
@pennae Do ypu mind running this PR in your perf benchmark ? Alternatively, do you mind sharing the benchmark and the way you run it ? I would really like to knwo the impact of this PR on perfs.

As a side note, we should really embed some benchmarks in CI builds, and report as a graph. over time.
I know sqlite has used ~qemu~ [cachegrind](https://www.sqlite.org/cpu.html) cycles count as a reliable benchmark metric independent of the CI machine and load.

**Comment by layus** [comment]
@pennae Thanks for your detailed explanation. Given your benchmark, I would say the OP gives pretty good confidence that this does not decrease performance, so it's all green for me.

@Ericson2314 Not that I have any commit right though :-).
I guess the only issue concerns adding "debug" parameters to core functions, but I started it :-D.

**Review by roberth — APPROVED** [review]

**Comment by layus** [comment]
@roberth Did we just forget to merge these changes ?

**Comment by layus** [comment]
@polykernel  I do not think the branch needs to be restored. All the information is still retained here.

**Pull request closed without merging**
